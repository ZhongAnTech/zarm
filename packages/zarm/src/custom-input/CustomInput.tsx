import * as React from 'react';
import classnames from 'classnames';
import { CloseCircleFill } from '@zarm-design/icons';
import KeyboardPicker from '../keyboard-picker';
import useClickAway from '../useClickAway';
import { getValue } from '../input/utils';
import { ConfigContext } from '../n-config-provider';
import type { BaseCustomInputProps } from './interface';

export type CustomInputProps = BaseCustomInputProps &
  Omit<React.InputHTMLAttributes<HTMLInputElement>, 'type' | 'onChange' | 'onFocus' | 'onBlur'>;

const CustomInput = React.forwardRef<unknown, CustomInputProps>((props, ref) => {
  const {
    type,
    clearable,
    readOnly,
    autoFocus,
    className,
    disabled,
    value,
    defaultValue,
    maxLength,
    onChange,
    onBlur,
    onFocus,
    placeholder,
    ...restProps
  } = props;

  const wrapperRef = (ref as any) || React.createRef<HTMLDivElement>();
  const contentRef = React.useRef<HTMLDivElement>(null);
  const pickerRef = React.useRef<HTMLDivElement>(null);
  const inputRef = React.useRef<HTMLInputElement>(null);
  const [currentValue, setCurrentValue] = React.useState(getValue({ value, defaultValue }, ''));
  const [focused, setFocused] = React.useState<boolean>(autoFocus!);
  const [area, setArea] = React.useState<React.RefObject<any>[]>([]);

  const showClearIcon =
    clearable &&
    typeof value !== 'undefined' &&
    currentValue.length > 0 &&
    typeof onChange === 'function';

  const { prefixCls: globalPrefixCls } = React.useContext(ConfigContext);
  const prefixCls = `${globalPrefixCls}-custom-input`;

  const cls = classnames(prefixCls, `${prefixCls}--${type}`, className, {
    [`${prefixCls}--disabled`]: disabled,
    [`${prefixCls}--focus`]: focused,
    [`${prefixCls}--clearable`]: showClearIcon,
    [`${prefixCls}--readonly`]: readOnly,
  });

  const onInputFocus = () => {
    if (disabled) return;

    setArea([contentRef, pickerRef]);
    setFocused(true);

    if (typeof onFocus === 'function') {
      onFocus(currentValue);
    }
  };

  const onInputBlur = () => {
    setFocused(false);

    if (typeof onBlur === 'function') {
      onBlur(currentValue);
    }
  };

  const onKeyClick = (key) => {
    if (['close', 'ok'].indexOf(key) > -1) {
      onInputBlur();
      return;
    }

    if (key !== 'delete' && currentValue.length >= maxLength!) {
      return;
    }

    const newValue =
      key === 'delete' ? currentValue.slice(0, currentValue.length - 1) : currentValue + key;

    if (typeof value === 'undefined') {
      setCurrentValue(newValue);
    }

    if (typeof onChange === 'function') {
      onChange(newValue);
    }
  };

  const onInputClear = (e) => {
    e.stopPropagation();
    setCurrentValue('');

    if (typeof onChange === 'function') {
      onChange('');
    }
  };

  const scrollToStart = () => {
    contentRef.current!.scrollLeft = 0;
  };

  const scrollToEnd = () => {
    contentRef.current!.scrollLeft = contentRef.current!.scrollWidth;
  };

  const focus = () => {
    onInputFocus();
  };

  const blur = () => {
    onInputBlur();
  };

  const renderClearIcon = showClearIcon && (
    <CloseCircleFill className={`${prefixCls}__clear`} onClick={onInputClear} />
  );

  const renderText = <div className={`${prefixCls}__content`}>{currentValue}</div>;

  const renderInput = (
    <div {...restProps} ref={wrapperRef} className={cls} onClick={onInputFocus}>
      <div className={`${prefixCls}__content`}>
        {(currentValue === undefined || currentValue === '') && !readOnly && (
          <div className={`${prefixCls}__placeholder`}>{placeholder}</div>
        )}
        <div ref={contentRef} className={`${prefixCls}__virtual-input`}>
          {currentValue}
        </div>
        <input ref={inputRef} type="hidden" value={currentValue} />
      </div>
      {renderClearIcon}
      <KeyboardPicker ref={pickerRef} visible={focused} type={type} onKeyClick={onKeyClick} />
    </div>
  );

  useClickAway(onInputBlur, area);

  React.useImperativeHandle(wrapperRef, () => ({
    focus,
    blur,
  }));

  React.useEffect(() => {
    setCurrentValue(getValue({ value, defaultValue }, ''));
  }, [value, defaultValue]);

  React.useEffect(() => {
    if (readOnly) return;

    if (focused) {
      scrollToEnd();
    } else {
      scrollToStart();
    }
  }, [readOnly, focused, currentValue]);

  return readOnly ? renderText : renderInput;
});

CustomInput.displayName = 'CustomInput';

CustomInput.defaultProps = {
  type: 'number',
  clearable: true,
  readOnly: false,
  autoFocus: false,
};

export default CustomInput;
