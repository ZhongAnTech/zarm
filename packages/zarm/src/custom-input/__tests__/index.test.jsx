import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import CustomInput from '../index';

describe('CustomInput', () => {
  it('renders correctly', () => {
    const wrapper = render(<CustomInput type="number" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders correctly if type isn't valid", () => {
    const wrapper = render(<CustomInput type="xxx" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('auto focus', () => {
    const id = String(Math.random());
    const wrapper = mount(<CustomInput id={id} autoFocus />);
    expect(wrapper.props().id).toEqual(id);
  });
});

describe('CustomInput.Number', () => {
  it('inputNumber value 0', () => {
    const wrapper = mount(<CustomInput clearable type="number" value={0} />);
    const input = wrapper.find('input[type="hidden"]');
    expect(input.instance().value).toEqual('0');
  });
});

describe('CustomInput.Price', () => {
  it('renders correctly', () => {
    const wrapper = mount(<CustomInput type="price" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('CustomInput.Idcard', () => {
  it('renders correctly', () => {
    const wrapper = mount(<CustomInput type="idcard" maxLength={18} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
