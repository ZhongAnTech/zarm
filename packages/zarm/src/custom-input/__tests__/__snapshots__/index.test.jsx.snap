// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomInput renders correctly 1`] = `
<div
  class="za-custom-input za-custom-input--number"
>
  <div
    class="za-custom-input__content"
  >
    <div
      class="za-custom-input__placeholder"
    />
    <div
      class="za-custom-input__virtual-input"
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`CustomInput renders correctly if type isn't valid 1`] = `
<div
  class="za-custom-input za-custom-input--xxx"
>
  <div
    class="za-custom-input__content"
  >
    <div
      class="za-custom-input__placeholder"
    />
    <div
      class="za-custom-input__virtual-input"
    />
    <input
      type="hidden"
      value=""
    />
  </div>
</div>
`;

exports[`CustomInput.Idcard renders correctly 1`] = `
<CustomInput
  autoFocus={false}
  clearable={false}
  disabled={false}
  maxLength={18}
  readOnly={false}
  type="idcard"
>
  <div
    className="za-custom-input za-custom-input--idcard"
    onClick={[Function]}
  >
    <div
      className="za-custom-input__content"
    >
      <div
        className="za-custom-input__placeholder"
      />
      <div
        className="za-custom-input__virtual-input"
      />
      <input
        type="hidden"
        value=""
      />
    </div>
    <KeyboardPicker
      destroy={true}
      onKeyClick={[Function]}
      type="idcard"
      visible={false}
    >
      <Popup
        destroy={true}
        lockScroll={true}
        mask={false}
        visible={false}
      />
    </KeyboardPicker>
  </div>
</CustomInput>
`;

exports[`CustomInput.Price renders correctly 1`] = `
<CustomInput
  autoFocus={false}
  clearable={false}
  disabled={false}
  readOnly={false}
  type="price"
>
  <div
    className="za-custom-input za-custom-input--price"
    onClick={[Function]}
  >
    <div
      className="za-custom-input__content"
    >
      <div
        className="za-custom-input__placeholder"
      />
      <div
        className="za-custom-input__virtual-input"
      />
      <input
        type="hidden"
        value=""
      />
    </div>
    <KeyboardPicker
      destroy={true}
      onKeyClick={[Function]}
      type="price"
      visible={false}
    >
      <Popup
        destroy={true}
        lockScroll={true}
        mask={false}
        visible={false}
      />
    </KeyboardPicker>
  </div>
</CustomInput>
`;
